### PR TITLE
blockstore: use word-level scans for completed data indexes

### PR DESCRIPTION
**Pull request by mjain-jump**
#### Problem
`update_completed_data_indexes` finds neighboring completed-data markers and checks contiguous shred ranges through iterator scans. That does unnecessary per-bit iterator work even though the data is stored in packed bit vectors.

#### Summary of Changes
Added fast previous/next set-bit scans to `BitVec` and semantic completed-index wrappers in `CompletedDataIndexes`. Added `ShredIndex::contains_range` and used these helpers in `update_completed_data_indexes` to reduce iterator work while preserving behavior.

#### Testing
Added edge-case and property tests comparing the new bit scans against the existing iterator behavior, plus a range-containment test for `ShredIndex`.

Fixes #

### CODE REVIEW

File: ledger/src/bit_vec.rs [hunk 1]
```diff
@@ -278,6 +279,109 @@ impl<const NUM_BITS: usize> BitVec<NUM_BITS> {
...
+        let last_word_idx = bound.checked_sub(1)? / BITS_PER_WORD;
+        let bit_in_word = (bound - 1) % BITS_PER_WORD;
+        let last_word =
+            self.words[last_word_idx] & (Word::MAX >> (BITS_PER_WORD - 1 - bit_in_word));
+        if last_word != 0 {
+            let msb = BITS_PER_WORD - 1 - last_word.leading_zeros() as usize;
+            return Some(last_word_idx * BITS_PER_WORD + msb);
+        }
+        // Scan full 8-byte groups before falling back to single bytes.
+        let mut chunks = self.words[..last_word_idx].rchunks_exact(WORDS_PER_U64);
```
Comment (cpubot) [inline ledger/src/bit_vec.rs]: Nit: could use `as_rchunks` to avoid the `chunk.try_into().expect()`
```rs
let (remainder, chunks) = self.words[..last_word_idx].as_rchunks::<WORDS_PER_U64>();
// ...
for &chunk in chunks.iter().rev() {
    let chunk_word = u64::from_le_bytes(chunk);
    // ...
```

File: ledger/src/bit_vec.rs [hunk 2]
```diff
@@ -278,6 +279,109 @@ impl<const NUM_BITS: usize> BitVec<NUM_BITS> {
...
+        let in_bounds = |pos: usize| (pos < NUM_BITS).then_some(pos);
+        let first_word_idx = from / BITS_PER_WORD;
+        let first_word = self.words[first_word_idx] & (Word::MAX << (from % BITS_PER_WORD));
+        if first_word != 0 {
+            return in_bounds(
+                first_word_idx * BITS_PER_WORD + first_word.trailing_zeros() as usize,
+            );
+        }
+        // Scan full 8-byte groups before falling back to single bytes.
+        let mut chunks = self.words[first_word_idx + 1..].chunks_exact(WORDS_PER_U64);
```
Comment (cpubot) [inline ledger/src/bit_vec.rs]: Similar nit to the above:
```rs
let (chunks, remainder) = self.words[first_word_idx + 1..].as_chunks::<WORDS_PER_U64>();
// ...
for &chunk in chunks {
    let chunk_word = u64::from_le_bytes(chunk);
    // ...
```

File: ledger/src/bit_vec.rs [hunk 3]
```diff
@@ -278,6 +279,109 @@ impl<const NUM_BITS: usize> BitVec<NUM_BITS> {
...
+    /// let mut bit_vec = BitVec::<64>::default();
+    /// bit_vec.insert(5);
+    /// assert_eq!(bit_vec.prev_set_bit(5), None);
+    /// assert_eq!(bit_vec.prev_set_bit(6), Some(5));
+    /// assert_eq!(bit_vec.prev_set_bit(64), Some(5));
+    /// ```
+    pub fn prev_set_bit(&self, bound: usize) -> Option<usize> {
+        let bound = bound.min(NUM_BITS);
+        let last_word_idx = bound.checked_sub(1)? / BITS_PER_WORD;
+        let bit_in_word = (bound - 1) % BITS_PER_WORD;
```
Comment (cpubot) [inline ledger/src/bit_vec.rs]: Can reuse `location_of`

```suggestion
        let (last_word_idx, bit_in_word) = location_of(bound.checked_sub(1)?);
```

File: ledger/src/bit_vec.rs [hunk 4]
```diff
@@ -278,6 +279,109 @@ impl<const NUM_BITS: usize> BitVec<NUM_BITS> {
...
+    /// ```
+    pub fn next_set_bit(&self, from: usize) -> Option<usize> {
+        if from >= NUM_BITS {
+            return None;
+        }
+        // Serialized data may contain non-zero tail bits past NUM_BITS. Match
+        // the range iterator by ignoring any hit outside the logical bit length.
+        let in_bounds = |pos: usize| (pos < NUM_BITS).then_some(pos);
+        let first_word_idx = from / BITS_PER_WORD;
+        let first_word = self.words[first_word_idx] & (Word::MAX << (from % BITS_PER_WORD));
```
Comment (cpubot) [inline ledger/src/bit_vec.rs]:
```suggestion
        let (first_word_idx, first_bit) = location_of(from);
        let first_word = self.words[first_word_idx] & (Word::MAX << first_bit);
```

**Comment by AshwinSekar** [comment]
Just wanted to point out that `disCA4efguFL6Wqa4pGdG7jpjC7C5uiKzKnhEBqchBe` will be activated on mainnet in ~6 epochs.

This restricts the completed data indices to align with FEC set boundaries (`shred_idx % 32 == 31`)

https://github.com/anza-xyz/agave/blob/8fc9840bf83a7afa6321ed13e2c5db4d2fc7f765/ledger/src/shred/filter.rs#L361-L367

Might make the code simpler/more performant as you don't need to scan the shreds in between

**Comment by mjain-jump** [comment]
> Just wanted to point out that `disCA4efguFL6Wqa4pGdG7jpjC7C5uiKzKnhEBqchBe` will be activated on mainnet in ~6 epochs.
> 
> This restricts the completed data indices to align with FEC set boundaries (`shred_idx % 32 == 31`)
> 
> https://github.com/anza-xyz/agave/blob/8fc9840bf83a7afa6321ed13e2c5db4d2fc7f765/ledger/src/shred/filter.rs#L361-L367
> 
> Might make the code simpler/more performant as you don't need to scan the shreds in between

can we keep this as-is for now and address this in a follow-up pr?

**Review by cpubot — APPROVED** [review]
LGTM

**Comment by mjain-jump** [comment]
> LGTM

Can someone please merge this for me? cc @AshwinSekar @cpubot

**Comment by steviez** [comment]
> Added fast previous/next set-bit scans to BitVec and semantic completed-index wrappers in CompletedDataIndexes

Out of curiosity, any idea on what kind of speedup this adds ?
